### PR TITLE
Provide port-specific fetch of submodules

### LINKED ISSUE #8036: Fetch submodules lazily
Instead of having to do a slow `make fetch-submodules` at the top level after cloning `circuitpython`, fetch the submodules lazily. Fetch the submodules for a particular port only when a build is done in that port directory.

... working on this.

### PR DESCRIPTION
**Pull request by dhalbert**
Fixes #8036.

- Rename top-level `make fetch-submodules` and `remove-submodules` to `fetch-all-submodules` and `remove-all-submodules`.
- In each `ports/*`, provide `make fetch-port-submodules`, which fetches only the submodules needed for that port.
- `make fetch-port-submodules` does not require `BOARD` to be set, but most port-specific Makefiles assumed that `BOARD` would always be set. They did not anticipate non-board-specific make targets. Fix that, by fixing how bad or missing `BOARD` is handled, and by guarding various actions in Makefiles to not do things if `BOARD` is not specified. New variables available now include `VALID_BOARD` and `ALL_PORT_BOARDS`.
- New script `tools/fetch-submodules.sh`. This could have been done solely in the makefiles, but it's less painful to use a shell script.

### CODE REVIEW

**Review by jepler — APPROVED** [review]
seems fine, didn't test. there's a low probability of folks invoking other targets in a port directory and not setting BOARD= but they can keep all the resulting pieces for themselves.

**Comment by dhalbert** [comment]
@jepler I found some bugs in the original intent of the PR -- I was so wrapped up in the other part I neglected to re-test. Tested again and `make fetch-port-submodules` now works.

File: py/circuitpy_defns.mk [hunk 1]
```diff
@@ -833,7 +833,8 @@ $(BUILD)/lib/libm/kf_rem_pio2.o: CFLAGS += -Wno-maybe-uninitialized
 # Fetch only submodules needed for this particular port.
 .PHONY: fetch-port-submodules
 fetch-port-submodules:
-	$(TOP)/tools/fetch-submodules.sh data extmod frozen lib $(CURDIR)
+	echo $(basename $(CURDIR))
```
Comment (jepler) [inline py/circuitpy_defns.mk:836]: maybe remove this echo?

**Review by jepler — CHANGES_REQUESTED** [review]

File: py/circuitpy_defns.mk (hunk 1 above)
Comment (dhalbert) [reply py/circuitpy_defns.mk]: Oops, yes, that was to prove that GNU Make basename is not the same /usr/bin/basename.

**Review by jepler — APPROVED** [review]
Thanks! again, didn't test.

**Pull request merged**